torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,467 @@
+
+    local_results = _scaled_dot_product_flash_attention(
+        op_info.mesh,
+        *op_info.local_args,  # type: ignore[arg-type]
+        **op_info.local_kwargs,  # type: ignore[arg-type]
+    )
+
+    return DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)
+
+
+def _merge_sdpa(
+    chunks: List[torch.Tensor], logsumexps: List[torch.Tensor]
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    This merges multiple scaled dot product attention chunks by using the
+    provided logsumexps to rescale the chunks before summing.
+
+    Args:
+        chunks (List[torch.Tensor]): A list of scaled dot product attention chunks
+        logsumexps (List[torch.Tensor]): A list of logsumexps for each chunk
+
+    Returns:
+        out (torch.Tensor): The merged scaled dot product attention
+        softmax_lse (torch.Tensor): The logsumexp of the merged scaled dot product attention
+    """
+    assert len(chunks) == len(logsumexps)
+
+    softmax_lse = torch.stack([lse.exp() for lse in logsumexps]).sum(dim=0).log_()
+
+    out = []
+    for i, (chunk, chunk_lse) in enumerate(zip(chunks, logsumexps)):
+        softmax_lse_corrected = torch.exp(chunk_lse - softmax_lse)
+        out_corrected = chunk * softmax_lse_corrected.unsqueeze(-1).to(chunk.dtype)
+        out.append(out_corrected)
+    out = torch.stack(out).sum(dim=0)
+
+    return out, softmax_lse
+
+
+def _scaled_dot_product_flash_attention(
+    mesh: DeviceMesh,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    return_debug_mask: bool = False,
+    *,
+    scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, ...]:
+    if return_debug_mask:
+        raise NotImplementedError("return_debug_mask is not supported yet")
+
+    if is_causal and (query.size(2) != key.size(2)):
+        raise NotImplementedError(
+            "is_causal requires the same query and context sequence lengths"
+        )
+
+    pg = mesh.get_group()
+    assert isinstance(pg, dist.ProcessGroup), "must be single dimension"
+    rank = dist.get_rank(pg)
+    size = dist.get_world_size(pg)
+
+    # rank 0 sends to rank 1, rank 1 sends to rank 2, ..., rank n-1 sends to rank 0
+    right_dsts = list(range(1, size)) + [0]
+
+    next_kv = None
+
+    chunks = []
+    logsumexps = []
+    for i in range(size):
+        # overlap communication with compute
+        if next_kv is not None:
+            next_kv = ft_c.wait_tensor(next_kv)
+            key = next_kv[: key.numel()].reshape(key.shape)
+            value = next_kv[key.numel() :].reshape(value.shape)
+
+        if i < (size - 1):
+            next_kv = torch.cat([key.flatten(), value.flatten()])
+            next_kv = ft_c.permute_tensor(next_kv, right_dsts, pg)
+
+        is_causal_behavior = _is_causal_behavior(
+            rank=rank, world_size=size, i=i, is_causal=is_causal
+        )
+
+        if is_causal_behavior != _CausalBehavior.SKIP:
```

Comment:
do we need to throw error if `is_causal_behavior == _CausalBehavior.SKIP`?

Reply:
SKIP is an expected state. With causal attention we're approximating a triangular matrix so if we have a chunk that's in the triangle where it's all zeros we don't have to call sdpa at all